### PR TITLE
cranelift: Align Scalar and SIMD shift semantics

### LINKED ISSUE #4424: Craneline vector shift count semantics does not match docs
In implementing SIMD support for s390x, the precise semantics of handling too-large shift counts for Cranelift vector shift instructions seemed unclear to me.   In the existing aarch64 and x86_64 targets, the Cranelift instructions are mapped directly to the native instructions, which happen to treat shift counts larger than the lane width by shifting everything off the end of the lane, i.e. resulting in an all-zero (or all-sign-bit-copies) lane element.

This behavior is actually explicitly verified in the simd-bitwise-run.clif runtest:
```
    v0 = iconst.i32 17 ; note that this will shift off the end of each lane
    v1 = vconst.i16x8 [1 2 4 8 16 32 64 128]
    v2 = ishl v1, v0
```

However, looking at the documentation, we have this statement:
```
        Integer shift left. Shift the bits in ``x`` towards the MSB by ``y``
        places. Shift in zero bits to the LSB.

        The shift amount is masked to the size of ``x``.

        When shifting a B-bits integer type, this instruction computes:

            s &:= y \pmod B,
            a &:= x \cdot 2^s \pmod{2^B}.
```
Now, admittedly, this does not explicitly refer to vector types either way.  However, usually the semantics for the vector operation matches the one for the scalar operation, so I would have expected that the shift count would be considered modulo the lane size.

And since the s390x vector shift instruction *does* implement that exact modulo semantics, I've simply implemented the Cranelift instruction directly with the native s390x instruction - only to see a failure in the runtest.

Looking at the Wasmtime as main user of Cranelift, the WebAssembly vector shifts are also specified to use modulo semantics.  And indeed, the WebAssembly to Cranelift mapping adds an explicit modulo operation there:
```
      Operator::I8x16Shl | Operator::I16x8Shl | Operator::I32x4Shl | Operator::I64x2Shl => {
            let (a, b) = state.pop2();
            let bitcast_a = optionally_bitcast_vector(a, type_of(op), builder);
            let bitwidth = i64::from(type_of(op).lane_bits());
            // The spec expects to shift with `b mod lanewidth`; so, e.g., for 16 bit lane-width
            // we do `b AND 15`; this means fewer instructions than `iconst + urem`.
            let b_mod_bitwidth = builder.ins().band_imm(b, bitwidth - 1);
            state.push1(builder.ins().ishl(bitcast_a, b_mod_bitwidth))
        }
```

It seems to me it would be preferable to have the Cranelift IR semantics match the WebAssembly semantics, remove that explicit modulo operation during the WebAssembly->Cranelift translation, and add it to those back ends that need it.   This would allow efficient code generation on s390x without affecting performance on the other platforms.

In any case, the documentation should be updated to explicitly define the Cranelift shift count semantics for vector types.

FYI @cfallin @afonso360

### PR DESCRIPTION
**Pull request by afonso360**
👋 Hey,

In #4424 @uweigand reported that we have different semantics for scalar and SIMD shifts, which is unusual in our instruction set.

This PR corrects the AArch64 and x86 backends to mask the shift amount before performing SIMD shifts. The s390x backend should already follow this behavior.

It also reorganizes the `ishl`/`ushl`/`sshl` tests into separate files. Since we had a bunch of duplicate tests in hard to find files that were somewhat unrelated.

I tried to keep changes contained in separate commits so that it's easier to review.

Fixes #4424
CC: @cfallin @abrown @uweigand @akirilov-arm

### CODE REVIEW

**Review by jameysharp — APPROVED** [review]
I appreciate the way you split up these commits! I wish I could ask CI to test after each commit, but it looks plausible to me that the tests would all pass at every step.

I'm not comfortable reviewing the aarch64 and x64 changes with any confidence. They look reasonable to me but I think others should review them to be sure.

I think your changes to the runtests look great, with a nice range of tests to cover this behavior. And the cranelift-wasm changes seem obviously correct.

Last thing I'm wondering is whether this will affect any other users of Cranelift, like @bjorn3.

File: cranelift/codegen/src/isa/aarch64/lower.isle
Comment (akirilov-arm) [inline cranelift/codegen/src/isa/aarch64/lower.isle:991]: Just a quick comment - I am surprised that passing `$I32` to `sub` works for `I64X2` vectors, since the shift amount should end up being a large positive number (or 0) instead of a small negative number.

File: cranelift/codegen/src/isa/aarch64/lower.isle
Comment (akirilov-arm) [inline cranelift/codegen/src/isa/aarch64/lower.isle:1041]: Ditto.

File: cranelift/filetests/filetests/isa/aarch64/arithmetic.clif
Comment (akirilov-arm) [inline cranelift/filetests/filetests/isa/aarch64/arithmetic.clif:345]: PR #4521 should remove these unnecessary instructions.

File: cranelift/codegen/src/isa/aarch64/lower.isle
Comment (afonso360) [reply cranelift/codegen/src/isa/aarch64/lower.isle]: I'm guessing `dup` will sign extend? That's the only way I can see this working, but I can't find any mention of sign extending in the docs.

I've switched to a 64 bit sub.

File: cranelift/filetests/filetests/isa/aarch64/arithmetic.clif
Comment (afonso360) [reply cranelift/filetests/filetests/isa/aarch64/arithmetic.clif:345]: 👍 Will rebase once that lands

File: cranelift/codegen/src/isa/aarch64/lower.isle
Comment (akirilov-arm) [reply cranelift/codegen/src/isa/aarch64/lower.isle]: Ah, now I remember why the original code works and is in fact correct, but I still like the usage of `$I64` - I forgot that the vector shifts used only **the least significant byte** of the shift amount, so there is in fact masking, just not the kind we need.

File: cranelift/codegen/src/isa/aarch64/lower/isle.rs
Comment (akirilov-arm) [inline cranelift/codegen/src/isa/aarch64/lower/isle.rs:338]: Could we add a `debug_assert!()` that `ty.lane_bits()` is a power of 2?

**Review by akirilov-arm — COMMENTED** [review]
The AArch64 changes look good to me, with a just a tiny nit.

**Review by akirilov-arm — APPROVED** [review]
Someone else should have a look at the x64 bits.

File: cranelift/codegen/src/isa/aarch64/lower/isle.rs
Comment (cfallin) [inline cranelift/codegen/src/isa/aarch64/lower/isle.rs:339]: There is a `.is_power_of_two()` method on the integer types that makes this clearer, IMHO (internally I suspect it uses exactly this idiom).

File: cranelift/codegen/src/isa/x64/lower.isle
Comment (cfallin) [inline cranelift/codegen/src/isa/x64/lower.isle:576]: This comment is slightly out of date now ("can each use a single instruction, once the shift amount is masked")

File: cranelift/codegen/src/isa/x64/lower.isle
Comment (cfallin) [inline cranelift/codegen/src/isa/x64/lower.isle:683]: Likewise here, let's update this comment

**Review by cfallin — APPROVED** [review]
Looks great to me! Just a few nits below. The final payoff in simpler code in the Wasm translator is great to see.

**Pull request merged**
